ral-facilities/datagateway-api
# Improve Client Handling

This PR will close #209 

## Description
This branch improves how client objects are managed across the API. Before this change, client objects were created towards the beginning of each request. These objects take several seconds to create, so caused the API to run very slowly. Clients are now managed by an LRU cache, which fetches clients from an object pool when needed. A pool of clients is created at startup, and are moved between the cache and back to the pool as required. I've documented how this works, so hopefully what I've put in there (and comments in the code) should be enough to not need to repeat all that in this PR. 

For reference, 74f1edd78b56b10618ec1c57f8016eb6b5073ad7 is the first commit where the cache-pool design appears and the commit message gives quite a bit of detail.

I've added tests for the LRU cache class that I've extended to ensure it's created with the configured cache size and that popping a client off the cache releases it back to the pool.

## Testing Instructions
Follow the documentation (https://github.com/ral-facilities/datagateway-api/blob/bugfix/client-cache-%23209/README.md#client-handling) to configure the cache and pool sizes. The login endpoint (`POST /sessions`) can be used to create session IDs, and any of the GET endpoints can be used to allocate a client object to that session ID. The combination of those two actions will allow the reviewer to test the cache and pool design, to ensure it works as intended when configured to correct proportions.

- [ ] Ensure the cache can grow to the configured size and that popping a client off the cache releases it back to the pool
- [ ] Ensure the pool creates the configured number of objects at startup (will be printed out to the console)
- [ ] Ensure the pool can grow over the initial size
- [ ] Test the e2e tests on DataGateway are happy with this change. My changes on `bugfix/casing-json-fields#94` aren't complete (some e2e tests fail), but running the DataGateway Download e2e tests should give a good representation (they pass happily on my machine using this branch)

## Agile Board Tracking
Connect to #209
